### DISCUSSION TITLE
How do I start and use the project?

### DISCUSSION THREAD

**Discussion by tonowie** [stustapay/stustapay#408] [Q&A] [answered]
hello,
I was hoping to start and test the application to see what exactly it does and whether it meets my requirements. I struggle to set-up an environment and use it though :(

Over the weekend I forked the project and created [compose, docker files and config](https://github.com/tonowie/stustapay/tree/compose/compose) to build and run the apps. Now everything is running, on the domain of my choosing (still to build the android app but the rest is mostly working) BUT I do not know how to use the system and I can find very little documentation. Can you please help me out?

1. administration and customerportal are running
2. how do I log-in to the administration? is there some default admin user? how do I create new user if not? What is the administration for?
3. what is the general workflow with the system? Is there some documentation I missed that describes how to use the system for some event and general capabilities it has? e.g. one of my reuirements is tracking of "one free beer" tokens for customers

**Comment by TheJJ** [comment, marked as answer]
hi! what are your requirements? :)

sure we can help you how to use the system!
you first have to create an admin user, you create it with `pdm run python3 -m stustapay -c etc/config.devel.yaml admin add-user --node-id 0` for the root node.

you should use this super-user only to create new events and admins for the new event (even for yourself).

the workflow is:
- customers always interact with the android app. the app is operated by workers on the event (entry, sales, ...)
- customer buys ticket via cash or card, gets a nft-tag in return (card or wristband)
- money can be topped up on the wristband by cash or card at top-up stations or online
- money can be payed out at a top-up stations as cash, to empty your leftover credits
- money can be payed out after the event to a bank account when people register their IBAN on the customerportal.
- money can be spent at food booths, bars, fun rides, ...
- people can be granted vouchers (as a reward for their helper shift, or as a make-good due to whatever problem)
- vouchers are automatically used when people buy stuff, since products can have voucher values.

i hope this helps you already? :smile_cat:

**Reply by tonowie** [reply]
Regarding my requirements the aim is to host cashless events and provide free vouchers (e.g. one beer per attendee). The idea was to use QR codes for purchases but I think it is negotiable and NFC is way better.

> you first have to create an admin user, you create it with `pdm run python3 -m stustapay -c etc/config.devel.yaml admin add-user --node-id 0` for the root node.

MAGIC, thank you!! :) This got me to the administration!!!

> the workflow is:
> ....

thank you very much for all this. I will have to spin up few more services and get the app going and I should be able to use the system end-to-end.

One more question though: What I have right now is docker image for the api services and each api service runs as separate container. Is that OK? Your approach suggests it is running on server with systemd services running for each api - so I am a bit afraid there might be some shared cache or something

**Reply by TheJJ** [reply]
state is only stored in the database, the services can run independently in seperate containers. we run the customer portal and api (terminal/admin) and database in different VMs.

**Reply by tonowie** [reply]
cool thank you. I will explore the system and try the app. It looks great though. good job

**Reply by tonowie** [reply]
@TheJJ I am able to use the system as admin now but I am struggling to get the terminal app working. After linking the app with terminal via the QR code I get error that I do not know how to interpret

> config error: bad response: Illegal input: Field 'detail' is required for type with serial name de.stustapay.libssp.net.parseDetailException.ExceptionDetail', but it was missing at path: $

That error is there for about 2 minutes while nginx access log shows repeated requests to config endpoint with 404 response. Then it changes to timeout to that endpoint with connect_timeout=5000ms

I can query the https://terminal.event.domain/apihealth endpoint in browser and my terminal server in config yaml is configured as follows

```
terminalserver:
  base_url: "https://terminal.event.domain/api"
  host: "terminalserver-api"
  port: 8083
```

Not really sure what to look for. all services are configured the same and have the same proxypass :(

**Reply by tonowie** [reply]
this is logged for requests by app
```
stustapay-terminalserver-api-1  | INFO:     172.18.0.10:36868 - "GET /api/config HTTP/1.1" 404 Not Found
stustapay-proxy-1               | 192.168.1.226 - - [20/Nov/2024:21:01:57 +0000] "GET /api/config HTTP/1.1" 404 120 "-" "Ktor client" "-"
```
this is logged for the request by browser but with no auth
```
stustapay-terminalserver-api-1  | INFO:     172.18.0.10:48860 - "GET /api/config HTTP/1.1" 401 Unauthorized
stustapay-proxy-1               | 192.168.1.167 - - [20/Nov/2024:21:03:11 +0000] "GET /api/config HTTP/1.1" 401 30 "-" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36" "-"
```
query to non-existent endpoint fails with 404 as expected
```
stustapay-terminalserver-api-1  | INFO:     172.18.0.10:44572 - "GET /api/badendpoint HTTP/1.1" 404 Not Found
stustapay-proxy-1               | 192.168.1.167 - - [20/Nov/2024:21:04:15 +0000] "GET /api/badendpoint HTTP/1.1" 404 22 "-" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130
```
healthcheck from browser is fine
```
stustapay-proxy-1               | 192.168.1.167 - - [20/Nov/2024:21:05:33 +0000] "GET /api/health HTTP/1.1" 200 20 "-" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36" "-"
stustapay-terminalserver-api-1  | INFO:     172.18.0.10:48830 - "GET /api/health HTTP/1.1" 200 OK
```

**Reply by tonowie** [reply]
semms like I have been chasing flawed error handling there :) raised #409
